### PR TITLE
Change ETag type

### PR DESCRIPTION
**Pull request by maririos**
Changes ETag from String to ETag type in ConfigurationSetting.

For the mocks, JSON.NET is not serializing ETag as needed. For example,
```
setting = new ConfigurationSetting()
{
   ...
  ETag = new ETag("pepito")
}

string json = JsonConvert.SerializeObject(setting ).ToLowerInvariant();
```
With this code, `json = "{etag:{}}"`. So I am overwriting it so that `json = {etag:pepito}`
but I am sure there must be a better way to do this ....

cc: @KrzysztofCwalina 

-----------------------
**Update**
I'm creating the `_responseContent` instead of leaving it to JSON.NET.
We still have a dependency on it for MockBatch that I will remove later in another PR. Issue https://github.com/Azure/azure-sdk-for-net/issues/5383

### CODE REVIEW

**Comment by jsquire** [comment]
To do this, you'll most likely want to create a custom converter to control the serialization.   That will let you intercept the read/write of the JSON value during serialization without having to manipulate it after.

The newer versions (12+) of JSON.NET have a nice little [helper class](https://www.newtonsoft.com/json/help/html/CustomJsonConverterGeneric.htm) to make this easy, but the [long way](https://www.newtonsoft.com/json/help/html/CustomJsonConverter.htm) isn't too bad either.

Once you've got a converter, you can just annotate the property that you want to apply it to in order to influence the serialization:

```c#
public sealed class ConfigurationSetting : IEquatable<ConfigurationSetting>
{
	/// <summary>
	/// An ETag indicating the state of a key-value within a configuration store.
	/// </summary>
	[JsonConverter(typeof(StringEtagConverter))]
	public ETag ETag { get; set; }
}
```

```c#
public class StringEtagConverter : Newtonsoft.Json.JsonConverter<ETag>
{
	public override void WriteJson(JsonWriter writer, ETag value, JsonSerializer serializer)
	{
		writer.WriteValue(value.ToString());
	}

	public override ETag ReadJson(JsonReader reader, Type objectType, ETag existingETag, bool hasExistingValue, JsonSerializer serializer)
	{
		string s = (string)reader.Value;

		return new ETag(s);
	}
}
```

If I understand what you're looking for, that should give you the JSON form you'd like.

```json
{
    "Key":"blah",
    "Label":null,
    "Value":"hi",
    "ContentType":null,
    "ETag":"some tag",
    "LastModified":"0001-01-01T00:00:00+00:00","Locked":false,"Tags":{}
}
```

**Comment by jsquire** [comment]
Forgive me if this was discussed previously, but with respect to the `ConfigurationSetting` class, i notice that you're overriding equality checks and the hash code generation, but haven't done overrides for the equality and inequality operators.     You may want to consider adding them.

**Comment by jsquire** [comment]
Forgive me, once again, for going out of the scope of this set of changes, but with respect to the `Tags` property implementation, I'm not sure that it is doing what you intended it do.

Perhaps I'm missing something, but the only `lock` that I see in the code is within the getter.  Because the setter is not synchronized, the null checks in the getter don't protect you from the value being set to null should a caller hit `set` right after the second null check condition.

Is there a reason that you chose not to just initialize the property at the time the class was instantiated?  Rather than paying the synchonization cost on each `get` accessor call, you may wish to spend a bit of memory to just initialize.

```c#
public IDictionary<string, string> Tags {
    get {
        if (_tags == null) {
            lock (_key) {
                if (_tags == null) {
                    _tags = new Dictionary<string, string>();
                }
            }
        }
        return _tags;
    }
    set  {
        _tags = value;
    }
}
````

**Comment by maririos** [comment]
>The newer versions (v12+) of Json.NET

@jsquire we actually want to delete the dependency on Json.NET. It is another issue, but it looks like this is a great opportunity to remove it all together. 
For the live tests, we have our own [serializer ](https://github.com/Azure/azure-sdk-for-net/blob/master/src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration/ConfigurationSettingParser.cs#L20). So I think the best way to go will be to use that one for the MockTests too.

I am looking into how to reuse it for both scenarios.

**Comment by maririos** [comment]
> I notice that you're overriding equality checks and the hash code generation, but haven't done overrides for the equality and inequality operators

Fair. Issue https://github.com/Azure/azure-sdk-for-net/issues/5475 opened

**Comment by maririos** [comment]
Updated description of issue with the solution for the serialization

File: src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs [hunk 1]
```diff
@@ -162,6 +144,7 @@ protected override void WriteResponseCore(HttpResponseMessage response)
                 bathItems.Items.Add(KeyValues[itemIndex++]);
             }
             string json = JsonConvert.SerializeObject(bathItems).ToLowerInvariant();
+            json = json.Replace("etag\":{}", "etag\":\"c3c231fd-39a0-4cb6-3237-4614474b92c1\"");
```
Comment (KrzysztofCwalina) [inline src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs]: @jamesnk, we know we can write a custom converter to serialize our ETag type, but maybe there is a better way. Custom  converter seems like a heavy hammer for something that amounts to struct wrapping a string. Is there some attribute or interface to tell JSON.NET to simply call ToString on this type when serializing?

Comment (JamesNK) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs]: I think a a TypeConverter that supports converting to a string will work.

Comment (maririos) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs]: @KrzysztofCwalina the plan is still to remove the dependency on JSON.NET, right?
Should we avoid adding more to it? I'm not addressing batches in this PR, but if the replacement I have for now is not enough, I could always do that and competitively eliminate the dependency

Comment (KrzysztofCwalina) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs]: I think your workaround is ok, if we will remove the dependency relatively soon. But I think it should not linger in the code as it's a bit of a hack :-)

Comment (maririos) [reply src/SDKs/Azure.ApplicationModel.Configuration/data-plane/Azure.ApplicationModel.Configuration.Tests/MockHttpClientTransport.cs]: lol. it is :( . Once I am done standardizing the API to match the other languages, I will work on this. (Issue #5383 )

**Review by KrzysztofCwalina — APPROVED** [review]

**Pull request merged**